src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+//! This module implements Nova's evaluation engine using multilinear KZG
+#![allow(non_snake_case)]
+use crate::{
+  errors::NovaError,
+  provider::{
+    kzg_commitment::KZGCommitmentEngine,
+    non_hiding_kzg::{KZGProverKey, KZGVerifierKey, UniversalKZGParam},
+    pedersen::Commitment,
+    traits::DlogGroup,
+  },
+  traits::{
+    commitment::{CommitmentEngineTrait, Len},
+    evaluation::EvaluationEngineTrait,
+    Engine as NovaEngine, Group, TranscriptEngineTrait, TranscriptReprTrait,
+  },
+};
+use core::marker::PhantomData;
+use ff::{Field, PrimeFieldBits};
+use group::{Curve, Group as _};
+use itertools::Itertools as _;
+use pairing::{Engine, MillerLoopResult, MultiMillerLoop};
+use rayon::prelude::*;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+/// Provides an implementation of a polynomial evaluation argument
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound(
+  serialize = "E::G1Affine: Serialize, E::Fr: Serialize",
+  deserialize = "E::G1Affine: Deserialize<'de>, E::Fr: Deserialize<'de>"
+))]
+pub struct EvaluationArgument<E: Engine> {
+  com: Vec<E::G1Affine>,
+  w: Vec<E::G1Affine>,
+  v: Vec<Vec<E::Fr>>,
+}
+
+/// Provides an implementation of a polynomial evaluation engine using KZG
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct EvaluationEngine<E, NE> {
+  _p: PhantomData<(E, NE)>,
+}
+
+impl<E, NE> EvaluationEngine<E, NE>
+where
+  E: Engine,
+  NE: NovaEngine<GE = E::G1, Scalar = E::Fr>,
+  E::G1: DlogGroup<PreprocessedGroupElement = E::G1Affine, Scalar = E::Fr>,
+  E::Fr: TranscriptReprTrait<E::G1>,
+  E::G1Affine: TranscriptReprTrait<E::G1>, // TODO: this bound on DlogGroup is really unusable!
+{
+  // This impl block defines helper functions that are not a part of
+  // EvaluationEngineTrait, but that we will use to implement the trait methods.
+  fn compute_challenge(
+    C: &E::G1Affine,
+    y: &E::Fr,
+    com: &[E::G1Affine],
+    transcript: &mut impl TranscriptEngineTrait<NE>,
+  ) -> E::Fr {
+    transcript.absorb(b"C", C);
+    transcript.absorb(b"y", y);
+    transcript.absorb(b"c", &com.to_vec().as_slice());
+
+    transcript.squeeze(b"c").unwrap()
+  }
+
+  // Compute challenge q = Hash(vk, C0, ..., C_{k-1}, u0, ...., u_{t-1},
+  // (f_i(u_j))_{i=0..k-1,j=0..t-1})
+  fn get_batch_challenge(
+    C: &[E::G1Affine],
+    u: &[E::Fr],
+    v: &[Vec<E::Fr>],
+    transcript: &mut impl TranscriptEngineTrait<NE>,
+  ) -> E::Fr {
+    transcript.absorb(b"C", &C.to_vec().as_slice());
```

Comment:
ditto. I'm interesting in upper bounds of expected sizes for `C`, `u`, `v` slices

Reply:
U is 3, v is 3*log(N)